pkg/forward/forward.go

```diff
@@ -334,7 +334,10 @@ func (isdf *InterStepDataForward) ackFromBuffer(ctx context.Context, offsets []i
 		Steps:    math.MaxInt,
 		Duration: time.Millisecond * 10,
 	}
-	var ackOffsets = offsets
+
+	// we only need to acknowledge every distinct offset once, for now, we add de-duplication logics here before sending offsets to fromBuffer.Ack().
+	// as of now, we don't force fromBuffer to implement Ack() as an idempotent function, but we should for future implementations.
+	var ackOffsets = isb.DeduplicateOffsets(offsets)
```

Comment:
This introduces extra O(n) time (and space) complexity for every batch. The time could be saved by moving the dedup logic to outside of ackFromBuffer.

Simply change the following code 

```
var readOffsets = make([]isb.Offset, len(readMessages))
for idx, m := range udfResults {
	readOffsets[idx] = m.readMessage.ReadOffset
}
err = isdf.ackFromBuffer(ctx, readOffsets)
```

to

```
set := make(map[string]bool)
var readOffsets []isb.Offset
for _, m := range udfResults {
	if ok := set[m.readMessage.ReadOffset.String()]; !ok {
		set[m.readMessage.ReadOffset.String()] = true
		readOffsets = append(readOffsets, m.readMessage.ReadOffset)
	}
}
```

But the O(n) space complexity for map still exist.

Reply:
jfyi, you can replace `bool` with `struct{}` that takes 0 space. we are just looking for the presence/absence of a key so the value shouldn't matter.